File: infitu/scripts/RewardAccumulator.py
#!/usr/bin/env python
import rospy
from threading import Lock
from percepto_msgs.msg import RewardStamped
from infitu.srv import SetRecording

class RewardAccumulator:
    """Listens to and sums or integrates a stream of rewards."""

    def __init__( self ):
        self.Reset()
        self.lock = Lock()

        self.default_on_empty = rospy.get_param( '~default_on_empty', False )
        self.default_value = float( rospy.get_param( '~default_value', 0 ) )

        self.time_integrate = rospy.get_param( '~mode/time_integrate', False )
        self.normalize_time = rospy.get_param( '~mode/normalize_time', False )
        self.normalize_count = rospy.get_param( '~mode/normalize_count', False )

        buff_size = rospy.get_param( '~buffer_size', 100 )
        self.sub = rospy.Subscriber( 'reward', RewardStamped, self.RewardCallback,
                                     queue_size = buff_size )

        self.recording_service = rospy.Service( '~set_recording', 
                                                SetRecording, 
                                                self.RecordCallback )

    def Reset( self ):
        self.acc = 0
        self.duration = 0
        self.count = 0
        self.last_reward = None

    def RewardCallback( self, msg ):
        with self.lock:
            if self.last_reward is None:
                self.last_reward = msg
            dt = ( msg.header.stamp - self.last_reward.header.stamp ).to_sec()
            self.count += 1
            self.duration += dt

            if self.time_integrate:
                self.acc += ( msg.reward + self.last_reward.reward ) * 0.5 * dt
            else:
                self.acc += msg.reward

            self.last_reward = msg

    def RecordCallback( self, req ):
        with self.lock:
            # If we are starting recording, reset state
            if req.enable_recording:
                self.Reset()
                return 0

            if self.default_on_empty and self.count == 0:
                return self.default_value

            # Else return the state
            evaluation = self.acc
            if self.normalize_time:
                if self.duration == 0:
                    evaluation = self.default_value
                else:
                    evaluation = evaluation / self.duration
            if self.normalize_count:
                if self.count == 0:
                    evaluation = self.default_value
                else:
                    evaluation = evaluation / self.count
            return evaluation

if __name__=="__main__":
    rospy.init_node( "reward_accumulator" )
    try:
        ra = RewardAccumulator()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: infitu/nodes/rewards/reward_accumulator.cpp
#include <ros/ros.h>

#include "percepto_msgs/RewardStamped.h"
#include "infitu/SetRecording.h"

#include "argus_utils/utils/ParamUtils.h"

using namespace argus;

class RewardAccumulator
{
public:

	RewardAccumulator( ros::NodeHandle& nh, ros::NodeHandle& ph )
		: _defaultValue( 0 )
	{
		GetParam( ph, "default_on_empty", _defaultOnEmpty, false );
		if( _defaultOnEmpty )
		{
			GetParamRequired( ph, "default_value", _defaultValue );
		}

		GetParam( ph, "time_integrate", _useIntegration, false );
		GetParam( ph, "normalize_by_time", _normalizeTime, false );
		GetParam( ph, "normalize_by_count", _normalizeCount, false );
		GetParam( ph, "max_dt", _maxDt, 1.0 );

		unsigned int buffSize;
		GetParam<unsigned int>( ph, "buffer_size", buffSize, 100 );
		_rewardSub = nh.subscribe( "reward", buffSize,
		                           &RewardAccumulator::RewardCallback, this );

		_recordServer = ph.advertiseService( "set_recording",
		                                     &RewardAccumulator::RecordingCallback,
		                                     this );
	}

private:

	void Reset()
	{
		_rewardAcc = 0;
		_rewardCount = 0;
		_rewardDur = 0;
		_lastRewardTime = ros::Time( 0 );
	}

	void RewardCallback( const percepto_msgs::RewardStamped::ConstPtr& msg )
	{
		// Indicate need for initialization with zero last time
		if( _lastRewardTime.isZero() )
		{
			_lastRewardTime = msg->header.stamp;
			_lastRewardValue = msg->reward;
			return;
		}

		double dt = (msg->header.stamp - _lastRewardTime).toSec();
		// Catch NaN/inf rewards and negative dts
		if( dt > _maxDt )
		{
			ROS_WARN_STREAM( "Received dt of " << dt << " larger than max " << _maxDt );
			Reset();
			return;
		}
		if( dt < 0 )
		{
			ROS_WARN_STREAM( "Received negative dt! Resetting..." );
			Reset();
			return;
		}

		if( std::isnan( msg->reward ) )
		{
			ROS_WARN_STREAM( "Received NaN reward!" );
			return;
		}

		double rewardInc = msg->reward;
		if( _useIntegration )
		{
			// Simple trapezoid rule
			rewardInc = 0.5 * (_lastRewardValue + msg->reward) * dt;
		}

		_rewardAcc += rewardInc;
		_rewardCount += 1;
		_rewardDur += dt;
		_lastRewardTime = msg->header.stamp;
		_lastRewardValue = msg->reward;
	}

	bool RecordingCallback( infitu::SetRecording::Request& req,
	                        infitu::SetRecording::Response& res )
	{
		if( req.enable_recording )
		{
			Reset();
			return true;
		}

		// Else return recording
		// Check for empty
		if( _rewardCount == 0 )
		{
			ROS_WARN_STREAM( "Received no rewards!" );
			res.evaluation =  _defaultOnEmpty ? _defaultValue :
			                 std::numeric_limits<double>::quiet_NaN();
			return true;
		}

		res.evaluation = _rewardAcc;
		if( _normalizeTime )
		{
			res.evaluation = res.evaluation / _rewardDur;
		}
		if( _normalizeCount )
		{
			res.evaluation = res.evaluation / _rewardCount;
		}
		return true;
	}

	double _rewardAcc;
	unsigned int _rewardCount;
	double _rewardDur;
	ros::Time _lastRewardTime;

	double _lastRewardValue;

	ros::Subscriber _rewardSub;
	ros::ServiceServer _recordServer;

	bool _defaultOnEmpty;
	double _defaultValue;

	double _maxDt;
	bool _useIntegration;
	bool _normalizeTime;
	bool _normalizeCount;
};

int main( int argc, char** argv )
{
	ros::init( argc, argv, "reward_accumulator" );
	ros::NodeHandle nh, ph( "~" );
	RewardAccumulator ra( nh, ph );
	ros::spin();
	return 0;
}